dotnet/src/Extensions/Planning.ActionPlanner/ActionPlanner.cs

```diff
@@ -0,0 +1,241 @@
+        }
+        catch (Exception e)
+        {
+            throw new PlanningException(PlanningException.ErrorCodes.InvalidPlan,
+                "Plan parsing error, invalid JSON", e);
+        }
+
+        if (planData == null)
+        {
+            throw new PlanningException(PlanningException.ErrorCodes.InvalidPlan, "The plan deserialized to a null object");
+        }
+
+        // Build and return plan
+        ISKFunction function;
+        if (planData.Plan.Function.Contains("."))
+        {
+            var parts = planData.Plan.Function.Split('.');
+            function = this._context.Skills!.GetFunction(parts[0], parts[1]);
+        }
+        else
+        {
+            function = this._context.Skills!.GetFunction(planData.Plan.Function);
+        }
+
+        var plan = new Plan(goal);
+        plan.AddSteps(function);
+
+        // Create a plan using the function and the parameters suggested by the planner
+        var variables = new ContextVariables();
+        foreach (KeyValuePair<string, string> p in planData.Plan.Parameters)
+        {
+            plan.State[p.Key] = p.Value;
+        }
+
+        Console.WriteLine(JsonSerializer.Serialize(planData, new JsonSerializerOptions { WriteIndented = true }));
+
+        var context = this._kernel.CreateNewContext();
+        context.Variables.Update(variables);
+
+        return plan;
+    }
+
+    // TODO: use goal to find relevant functions in a skill store
+    /// <summary>
+    /// Native function returning a list of all the functions in the current context,
+    /// excluding functions in the planner itself.
+    /// </summary>
+    /// <param name="goal">Currently unused. Will be used to handle long lists of functions.</param>
+    /// <param name="context">Function execution context</param>
+    /// <returns>List of functions, formatted accordingly to the prompt</returns>
+    [SKFunction("List all functions available in the kernel")]
+    [SKFunctionName("ListOfFunctions")]
+    [SKFunctionInput(Description = "The current goal processed by the planner", DefaultValue = "")]
+    public string ListOfFunctions(string goal, SKContext context)
+    {
+        Verify.NotNull(context.Skills, "The planner requires a non-null skill collection");
+        var functionsAvailable = context.Skills.GetFunctionsView();
+
+        // Prepare list using the format used by skprompt.txt
+        var list = new StringBuilder();
+        this.PopulateList(list, functionsAvailable.NativeFunctions);
+        this.PopulateList(list, functionsAvailable.SemanticFunctions);
+
+        return list.ToString();
+    }
+
+    // TODO: generate string programmatically
+    // TODO: use goal to find relevant examples
+    [SKFunction("List a few good examples of plans to generate")]
+    [SKFunctionName("GoodExamples")]
+    [SKFunctionInput(Description = "The current goal processed by the planner", DefaultValue = "")]
+    public string GoodExamples(string goal, SKContext context)
```

Comment:
Why an SKFunction? Is the AI meant to call this?

Reply:
yes see the prompt